op-batcher/batcher/driver.go

```diff
@@ -387,6 +388,47 @@ func (l *BatchSubmitter) setTxPoolState(txPoolState TxPoolState, txPoolBlockedBl
 	l.txpoolMutex.Unlock()
 }
 
+// syncAndPublish computes actions to take based on the current sync status, and then executes
+// them, updating the channel manager and sending transactions to the DA layer.
+func (l *BatchSubmitter) syncAndPublish(syncStatus *eth.SyncStatus, queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], daGroup *errgroup.Group) {
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
+
+	blocks := l.channelMgr.blocks
+	channels := l.channelMgr.channelQueue
+	// Decide appropriate actions
+	syncActions, outOfSync := computeSyncActions(*syncStatus, l.prevCurrentL1, blocks, channels, l.Log)
+
+	if outOfSync {
+		// If the sequencer is out of sync
+		// do nothing and wait to see if it has
+		// got in sync on the next tick.
+		l.Log.Warn("Sequencer is out of sync, retrying next tick.")
+		return
+	}
+
+	l.prevCurrentL1 = syncStatus.CurrentL1
+
+	// Manage existing state / garbage collection
+	if syncActions.clearState != nil {
+		l.channelMgr.Clear(*syncActions.clearState)
+	} else {
+		l.channelMgr.pruneSafeBlocks(syncActions.blocksToPrune)
+		l.channelMgr.pruneChannels(syncActions.channelsToPrune)
+	}
+
+	if syncActions.blocksToLoad != nil {
+		// Get fresh unsafe blocks
+		if err := l.loadBlocksIntoState(l.shutdownCtx, syncActions.blocksToLoad.start, syncActions.blocksToLoad.end); errors.Is(err, ErrReorg) {
+			l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
+			l.waitNodeSyncAndClearState()
+			return
+		}
+	}
```

Comment:
We're now holding the mutex while loading blocks, which can take time, especially at startup. This is a significant change in locking behavior. I think it's fine, but we should have an eye on it.

The reorg case might be most concerning, since the `waitNodeSyncAndClearState` may take some time.

Reply:
Yes, a valid concern. I think we have to prefer atomicity to "performance" here. The main loop is definitely the most important now, since the receipts handling loop only deals with quite edge cases which the main loop would eventually catch. The throttling loop could be starved for longer than before, though, which could be a concern.

Reply:
In the end the flakiness of the test with this "long lock holding" approach was a signal that we aren't ready to make that change. So I have reverted to a pattern which releases the lock more often, between pruning and loading blocks (and in fact between each block that is loaded).